File: Code/Arrays_Bit_Manipulation/Subarray_Sum_Divisible_By_K.php
<?php

class Solution {
    function subarraysDivByK($nums, $k) {
        $remainderCount = [];
        $remainderCount[0] = 1;

        $prefixSum = 0;
        $count = 0;

        foreach ($nums as $num) {
            $prefixSum += $num;
            $remainder = (($prefixSum % $k) + $k) % $k;

            if (isset($remainderCount[$remainder])) {
                $count += $remainderCount[$remainder];
            }
            $remainderCount[$remainder] = ($remainderCount[$remainder] ?? 0) + 1;
        }

        return $count;
    }
}

// Test
$sol = new Solution();
$nums = [4, 5, 0, -2, -3, 1];
$k = 5;
$result = $sol->subarraysDivByK($nums, $k);
echo "Number of subarrays divisible by $k is: $result\n";

File: Code/Arrays_Bit_Manipulation/Subarray_Sum_Divisible_By_K.py
from collections import defaultdict

class Solution:
    def subarraysDivByK(self, nums: list[int], k: int) -> int:
        remainder_count = defaultdict(int)
        remainder_count[0] = 1

        prefix_sum = 0
        count = 0

        for num in nums:
            prefix_sum += num
            remainder = (prefix_sum % k + k) % k
            count += remainder_count[remainder]
            remainder_count[remainder] += 1

        return count

# Test
if __name__ == "__main__":
    sol = Solution()
    nums = [4, 5, 0, -2, -3, 1]
    k = 5
    print("Number of subarrays divisible by", k, "is:", sol.subarraysDivByK(nums, k))
